src/control/lib/control/server_meta.go

```diff
@@ -143,6 +166,199 @@ func SmdQuery(ctx context.Context, rpcClient UnaryInvoker, req *SmdQueryReq) (*S
 		return ctlpb.NewCtlSvcClient(conn).SmdQuery(ctx, pbReq)
 	})
 
+	ur, err := rpcClient.InvokeUnaryRPC(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	sr := new(SmdResp)
+	for _, hostResp := range ur.Responses {
+		if hostResp.Error != nil {
+			if err := sr.addHostError(hostResp.Addr, hostResp.Error); err != nil {
+				return nil, err
+			}
+			continue
+		}
+
+		if err := sr.addHostQueryResponse(hostResp, req.FaultyDevsOnly); err != nil {
+			return nil, err
+		}
+	}
+
+	return sr, nil
+}
+
+func (sr *SmdResp) addHostManageResponse(hr *HostResponse) error {
+	pbResp, ok := hr.Message.(*ctlpb.SmdManageResp)
+	if !ok {
+		return errors.Errorf("unable to unpack message: %+v", hr.Message)
+	}
+
+	hs := &HostStorage{
+		SmdInfo: &SmdInfo{},
+	}
+	for _, rResp := range pbResp.GetRanks() {
+		rank := ranklist.Rank(rResp.Rank)
+
+		for _, pbResult := range rResp.GetResults() {
+			sd := new(storage.SmdDevice)
+			if err := convert.Types(pbResult.Device, sd); err != nil {
+				return errors.Wrapf(err, "converting %T to %T", pbResult.Device, sd)
+			}
+			sd.Rank = rank
+			hs.SmdInfo.Devices = append(hs.SmdInfo.Devices, sd)
+		}
+	}
+
+	if sr.HostStorage == nil {
+		sr.HostStorage = make(HostStorageMap)
+	}
+	if err := sr.HostStorage.Add(hr.Addr, hs); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (sr *SmdResp) getHostManageRespErr(hr *HostResponse) error {
+	pbResp, ok := hr.Message.(*ctlpb.SmdManageResp)
+	if !ok {
+		return errors.Errorf("unable to unpack message: %+v", hr.Message)
+	}
+
+	var errMsgs []string
+	for _, rResp := range pbResp.GetRanks() {
+		for _, pbResult := range rResp.GetResults() {
+			if pbResult.Status == 0 {
+				continue
+			}
+			var id string
+			if pbResult.Device != nil {
+				id = pbResult.Device.Uuid
+				if id == "" {
+					id = pbResult.Device.TrAddr
+				}
+				id += " "
+			}
+			errMsgs = append(errMsgs, fmt.Sprintf("rank %d: %s%s", rResp.Rank, id,
+				daos.Status(pbResult.Status)))
+		}
+	}
+
+	if len(errMsgs) > 0 {
+		return errors.New(strings.Join(errMsgs, ", "))
+	}
+
+	return nil
+}
+
+// Pack SmdManage proto request oneof field.
+func packPBSmdManageReq(req *SmdManageReq, pbReq *ctlpb.SmdManageReq) error {
+	var optParsed bool
+
+	if req.SetFaulty {
+		optParsed = true
+		// Expect single UUID in request IDs field.
+		if err := checkUUID(req.IDs); err != nil {
+			return errors.Wrap(err, "bad device UUID to set-faulty")
+		}
+		pbReq.Op = &ctlpb.SmdManageReq_Faulty{
+			Faulty: &ctlpb.SetFaultyReq{
+				Uuid: req.IDs,
+			},
+		}
+	}
+
+	if req.ReplaceUUID != "" {
+		if optParsed {
+			return errSmdManageMultiOpsInReq
+		}
```

Comment:
suggestion - If we're only allowed to do one at a time, maybe defining opcodes would be a better model.